dangerzone/conversion/pixels_to_pdf.py

```diff
@@ -112,27 +84,39 @@ async def convert(
                     timeout=timeout,
                 )
 
+            # remove PNG file when it is no longer needed
+            os.remove(png_filename)
+
             self.percentage += percentage_per_page
 
         # Next operations apply to the all the pages, so we need to recalculate the
         # timeout.
         timeout = self.calculate_timeout(total_size, num_pages)
 
         # Merge pages into a single PDF
+        timeout_per_batch = get_batch_timeout(timeout, num_pages)
         self.update_progress(f"Merging {num_pages} pages into a single PDF")
-        args = ["pdfunite"]
-        for page in range(1, num_pages + 1):
-            args.append(f"{tempdir}/page-{page}.pdf")
-        args.append(f"{tempdir}/safe-output.pdf")
-        await self.run_command(
-            args,
-            error_message="Merging pages into a single PDF failed",
-            timeout_message=(
-                "Error merging pages into a single PDF, pdfunite timed out after"
-                f" {timeout} seconds"
-            ),
-            timeout=timeout,
-        )
+        for first_page, last_page in batch_iterator(num_pages):
```

Comment:
Question: what do we gain by batching the `pdfunite` operations?

At this point, we already have all the pages as PDF files on disk, and that is the main space hog, right? I get that with a single `pdfunite` call we need double the space, but that's a requirement for the conversion step as well, which takes the original PDF and creates a compressed one, and they both co-exist in the disk.

I have another suggestion here: using PyPDF2 instead, and perform compression online:

```python
from pypdf import PdfReader, PdfWriter

reader = PdfReader("example.pdf")
writer = PdfWriter()

for page in reader.pages:
    writer.add_page(page)

for page in writer.pages:
    # ⚠️ This has to be done on the writer, not the reader!
    page.compress_content_streams()  # This is CPU intensive!

with open("out.pdf", "wb") as f:
    writer.write(f)
 ```
 
 _(snippet taken from [the docs](https://pypdf.readthedocs.io/en/latest/user/file-size.html#lossless-compression))_
 
With PyPDF2, we can open each PDF, compress it, and immediately create the output PDF, without `pdfunite` and `ps2pdf`. We can even delete the underlying PDFs as we process them. The good thing about `PyPDF2` is that it's available in Fedora repos as well, so it will work on Qubes.

Reply:
> Question: what do we gain by batching the pdfunite operations?

It's [in the commit message](https://github.com/freedomofpress/dangerzone/pull/619/commits/f7190e3876ce3ab33fb0db9b6e94dc9ff872a531). So it's not size-related but rather ulimit related. We could change the ulimit, but since I had already done the batch logic, I figured it could be reused here.

Reply:
Oh snap. I didn't expect that Qubes had such a low ulimit (nor that `pdfunite` opens the files at the same time, jeez...).